File: src/bootme/installer.py
"""Installer functions for copying AHK files."""

import shutil
from pathlib import Path
from typing import List, Dict
from dataclasses import dataclass

from .utils import get_autohotkeys_dir


@dataclass
class InstallResult:
    """Result of installation operation."""
    installed: List[str]
    failed: List[str]
    skipped: List[str]


def ensure_dir(dir_path: str) -> None:
    """Ensure a directory exists, create if not.

    Args:
        dir_path: Directory path
    """
    Path(dir_path).mkdir(parents=True, exist_ok=True)


def install_ahk_file(file_name: str, target_dir: str) -> bool:
    """Install a single AHK file to the target directory.

    Args:
        file_name: Name of the file to install
        target_dir: Target installation directory

    Returns:
        bool: Success status
    """
    source_dir = get_autohotkeys_dir()
    source_path = source_dir.joinpath(file_name)
    target_path = Path(target_dir) / file_name

    try:
        ensure_dir(target_dir)
        # Read from package resource and write to target
        content = source_path.read_bytes()
        target_path.write_bytes(content)
        return True
    except Exception as e:
        print(f"Failed to install {file_name}: {e}")
        return False


def install_ahk_files(
    files: List[str],
    target_dir: str,
    skip_files: List[str] = None
) -> InstallResult:
    """Install multiple AHK files to the target directory.

    Args:
        files: Array of file names to install
        target_dir: Target installation directory
        skip_files: Array of file names to skip

    Returns:
        InstallResult: Object with installed, failed, and skipped lists
    """
    if skip_files is None:
        skip_files = []

    result = InstallResult(installed=[], failed=[], skipped=[])
    skip_set = set(skip_files)

    for file in files:
        if file in skip_set:
            result.skipped.append(file)
            continue

        success = install_ahk_file(file, target_dir)
        if success:
            result.installed.append(file)
        else:
            result.failed.append(file)

    return result

File: lib/installer.js
import { mkdirSync, copyFileSync, existsSync } from 'fs';
import { join } from 'path';
import { getAutohotkeysDir } from './utils.js';

/**
 * Ensure a directory exists, create if not
 * @param {string} dir - Directory path
 */
export function ensureDir(dir) {
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true });
  }
}

/**
 * Install a single AHK file to the target directory
 * @param {string} fileName - Name of the file to install
 * @param {string} targetDir - Target installation directory
 * @returns {boolean} Success status
 */
export function installAhkFile(fileName, targetDir) {
  const sourceDir = getAutohotkeysDir();
  const sourcePath = join(sourceDir, fileName);
  const targetPath = join(targetDir, fileName);

  try {
    ensureDir(targetDir);
    copyFileSync(sourcePath, targetPath);
    return true;
  } catch (error) {
    console.error(`Failed to install ${fileName}: ${error.message}`);
    return false;
  }
}

/**
 * Install multiple AHK files to the target directory
 * @param {string[]} files - Array of file names to install
 * @param {string} targetDir - Target installation directory
 * @param {string[]} skipFiles - Array of file names to skip
 * @returns {{ installed: string[], failed: string[], skipped: string[] }}
 */
export function installAhkFiles(files, targetDir, skipFiles = []) {
  const result = {
    installed: [],
    failed: [],
    skipped: []
  };

  const skipSet = new Set(skipFiles);

  for (const file of files) {
    if (skipSet.has(file)) {
      result.skipped.push(file);
      continue;
    }

    const success = installAhkFile(file, targetDir);
    if (success) {
      result.installed.push(file);
    } else {
      result.failed.push(file);
    }
  }

  return result;
}
